### PR TITLE
feat(condo): DOMA-9935 add minimum amount settings to acquiring integration

### PR DESCRIPTION
**Pull request by vovaaxeapolla**
Different Acquring Integrations can have different minimum payment amount limit. As we are going to implement partial payments for receipts, we need to check and to make prompts to user about minimum payment sum

### CODE REVIEW

**Review by dkoviazin — APPROVED** [review]

File: apps/condo/lang/en/en.json [hunk 1]
```diff
@@ -3134,6 +3134,7 @@
   "acquiringReceipt.invoice.row.count": "Count",
   "acquiringReceipt.invoice.row.vatPercent": "VAT",
   "acquiringReceipt.invoice.row.amount": "Amount",
-  "api.acquiring.multiPayment.error.receiptsWithInvoices": "Receipts and invoices are forbidden to be together"
+  "api.acquiring.multiPayment.error.receiptsWithInvoices": "Receipts and invoices are forbidden to be together",
+  "api.acquiring.payment.error.paymentAmountLessThanMinimum": "The payment amount must be greater than {minimumPaymentAmount}"
```
Comment (toplenboren) [inline apps/condo/lang/en/en.json:3138]:
```suggestion
  "api.acquiring.payment.error.paymentAmountLessThanMinimum": "Payment amount must be greater than {minimumPaymentAmount}"
```

File: apps/condo/domains/acquiring/constants/errors.js [hunk 2]
```diff
@@ -103,6 +105,12 @@ const GQL_ERRORS = {
         type: ERROR_TAX_REGIME_AND_VAT_NOT_MATCHED,
         message: 'Tax regime and vat values are not matched',
     },
+    PAYMENT_AMOUNT_LESS_THAN_MINIMUM: {
+        code: BAD_USER_INPUT,
+        type: WRONG_VALUE,
+        message: 'The payment amount is less than the minimum amount from acquiring integration',
```
Comment (toplenboren) [inline apps/condo/domains/acquiring/constants/errors.js:111]:
```suggestion
        message: 'Payment amount is less than the minimum amount from acquiring integration',
```

File: apps/condo/domains/acquiring/schema/AcquiringIntegration.js [hunk 3]
```diff
@@ -75,7 +75,10 @@ const AcquiringIntegration = new GQLListSchema('AcquiringIntegration', {
                 },
             },
         },
-
+        minimumPaymentAmount: {
+            ...POSITIVE_MONEY_AMOUNT_FIELD,
+            schemaDoc: 'The minimum payment amount that can be accepted',
```
Comment (toplenboren) [inline apps/condo/domains/acquiring/schema/AcquiringIntegration.js:80]:
```suggestion
            schemaDoc: 'Minimum payment amount that can be accepted',
```

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.js [hunk 4]
```diff
@@ -232,6 +234,19 @@ const RegisterMultiPaymentForInvoicesService = new GQLCustomSchema('RegisterMult
                     implicitFee: Big('0.0'),
                 })
 
+                const amountToPay = Big(totalAmount.amountWithoutExplicitFee)
+                    .add(Big(totalAmount.explicitServiceCharge))
+                    .add(Big(totalAmount.explicitFee))
+                if (acquiringIntegration.minimumPaymentAmount && Big(amountToPay).lt(acquiringIntegration.minimumPaymentAmount)) {
+                    throw new GQLError({
+                        ...ERRORS.PAYMENT_AMOUNT_LESS_THAN_MINIMUM,
+                        messageInterpolation: { minimumPaymentAmount: Big(acquiringIntegration.minimumPaymentAmount).toFixed(2) },
```
Comment (toplenboren) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.js:243]: While it is true that `2` is applicable for RUB, USD, EUR, and many other currencies we should support custom currencies, like BTC (why not?)

You can get radix for your current locale in acquiring integration

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.test.js [hunk 5]
```diff
@@ -277,4 +282,49 @@ describe('RegisterMultiPaymentForInvoicesService', () => {
...
+            })
+        })
+
+        afterEach(async () => {
+            await utils.updateAcquiringIntegration({ minimumPaymentAmount: null })
+        })
+
+        describe('Check minimum payment amount from acquiring integration', () => {
+
+            test('No limits for payment if no settings for minimumPaymentAmount in acquiring integration', async () => {
```
Comment (toplenboren) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.test.js:309]: Maybe.. payment for acquiring with minimumPaymentAmount = null ?

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.test.js [hunk 6]
```diff
@@ -277,4 +282,49 @@ describe('RegisterMultiPaymentForInvoicesService', () => {
...
+                const [invoice] = await createTestInvoice(utils.clients.admin, utils.organization, { status: INVOICE_STATUS_PUBLISHED })
+                const [result] = await registerMultiPaymentForInvoicesByTestClient(utils.clients.admin, { invoices: [{ id: invoice.id }] })
+                expect(result).toHaveProperty('multiPaymentId')
+            })
+
+            test('The payment amount is less than the minimum amount from acquiring integration', async () => {
```
Comment (toplenboren) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.test.js:315]: You also should add at least one positive test and one test where minimumPaymentAmount equals amountToPay

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentForVirtualReceiptService.js [hunk 7]
```diff
@@ -185,6 +189,17 @@ const RegisterMultiPaymentForVirtualReceiptService = new GQLCustomSchema('Regist
...
+                const amountToPay = Big(amount)
+                    .add(Big(paymentCommissionFields.explicitServiceCharge))
+                    .add(Big(paymentCommissionFields.explicitFee))
+                if (acquiringIntegration.minimumPaymentAmount && Big(amountToPay).lt(acquiringIntegration.minimumPaymentAmount)) {
+                    throw new GQLError({
+                        ...ERRORS.PAYMENT_AMOUNT_LESS_THAN_MINIMUM,
+                        messageInterpolation: { minimumPaymentAmount: Big(acquiringIntegration.minimumPaymentAmount).toFixed(2) },
+                    }, context)
+                }
+
```
Comment (toplenboren) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentForVirtualReceiptService.js:202]: Looks like this logic is common for multiple modules, do you think we can move this away to some sort of utility function or up the callstack?

Comment (toplenboren) [reply apps/condo/domains/acquiring/schema/RegisterMultiPaymentForVirtualReceiptService.js:202]: Maybe to the hooks of the `MultiPayment` model

File: apps/condo/lang/ru/ru.json [hunk 8]
```diff
@@ -3134,6 +3134,7 @@
   "acquiringReceipt.invoice.row.count": "Кол-во",
   "acquiringReceipt.invoice.row.vatPercent": "НДС",
   "acquiringReceipt.invoice.row.amount": "Стоимость",
-  "api.acquiring.multiPayment.error.receiptsWithInvoices": "Квитанции и счета не могут быть в одном платеже, увы. Но мы над этим работаем."
+  "api.acquiring.multiPayment.error.receiptsWithInvoices": "Квитанции и счета не могут быть в одном платеже, увы. Но мы над этим работаем.",
```
Comment (toplenboren) [inline apps/condo/lang/ru/ru.json:3137]:
```suggestion
  "api.acquiring.multiPayment.error.receiptsWithInvoices": "Квитанции и счета не могут быть в одном платеже.",
```

Comment (toplenboren) [reply apps/condo/lang/ru/ru.json:3137]: Know this is not your code

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js [hunk 9]
```diff
@@ -351,7 +354,7 @@ const RegisterMultiPaymentService = new GQLCustomSchema('RegisterMultiPaymentSer
                     if (distributionReceiptsIds.length !== uniqueDistributionReceiptsIds.size) {
                         throw new GQLError(ERRORS.BAD_AMOUNT_DISTRIBUTION_FOR_RECEIPTS, context)
                     }
-                    if (distributionReceiptsAmounts.filter(amount => amount.lte(MINIMUM_PAYMENT_AMOUNT)).length > 0) {
+                    if (distributionReceiptsAmounts.filter(amount => amount.lte(0)).length > 0) {
```
Comment (toplenboren) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js:357]: why `lte` here, when in other modules you throw error on `lt` ?

Comment (toplenboren) [reply apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js:357]: + why 0 here?

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js [hunk 10]
```diff
@@ -705,6 +696,25 @@ const RegisterMultiPaymentService = new GQLCustomSchema('RegisterMultiPaymentSer
...
+
+                const amountToPay = Big(totalAmount.amountWithoutExplicitFee)
+                    .add(Big(totalAmount.explicitServiceCharge))
+                    .add(Big(totalAmount.explicitFee))
+                if (acquiringIntegration.minimumPaymentAmount && Big(amountToPay).lt(acquiringIntegration.minimumPaymentAmount)) {
+                    throw new GQLError({
+                        ...ERRORS.PAYMENT_AMOUNT_LESS_THAN_MINIMUM,
+                        messageInterpolation: { minimumPaymentAmount: Big(acquiringIntegration.minimumPaymentAmount).toFixed(2) },
+                    }, context)
+                }
```
Comment (toplenboren) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js:708]: again

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.js (hunk 4 above)
Comment (VKislov) [reply apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.js:243]: may be write TODO and move it to backlog

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.js [hunk 11]
```diff
@@ -232,6 +234,19 @@ const RegisterMultiPaymentForInvoicesService = new GQLCustomSchema('RegisterMult
...
+                    }, context)
+                }
+
+                const payments = await Promise.all(paymentCreateInputs.map((paymentInput) => Payment.create(context, paymentInput)))
```
Comment (VKislov) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.js:247]: it seems the create operation is asynchronous. So without await there will be an array of promises in the pending state
```suggestion
                const payments = await Promise.all(paymentCreateInputs.map(async (paymentInput) => await Payment.create(context, paymentInput)))
```

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js [hunk 12]
```diff
@@ -553,88 +556,83 @@ const RegisterMultiPaymentService = new GQLCustomSchema('RegisterMultiPaymentSer
...
-
                     // All acquiring contexts must be finished
                     if (acquiringContexts.some(({ invoiceStatus }) => invoiceStatus !== CONTEXT_FINISHED_STATUS)) {
                         throw new GQLError(ERRORS.INVOICE_CONTEXT_NOT_FINISHED, context)
                     }
-
                     if (
                         !!billingIntegrationCurrencyCode && !!invoiceIntegrationCurrencyCode
-                        && billingIntegrationCurrencyCode !== invoiceIntegrationCurrencyCode
+                            && billingIntegrationCurrencyCode !== invoiceIntegrationCurrencyCode
```
Comment (VKislov) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js:635]:
```suggestion
                        && billingIntegrationCurrencyCode !== invoiceIntegrationCurrencyCode
```

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js [hunk 13]
```diff
@@ -705,6 +696,25 @@ const RegisterMultiPaymentService = new GQLCustomSchema('RegisterMultiPaymentSer
...
+                    .add(Big(totalAmount.explicitServiceCharge))
+                    .add(Big(totalAmount.explicitFee))
+                if (acquiringIntegration.minimumPaymentAmount && Big(amountToPay).lt(acquiringIntegration.minimumPaymentAmount)) {
+                    throw new GQLError({
+                        ...ERRORS.PAYMENT_AMOUNT_LESS_THAN_MINIMUM,
+                        messageInterpolation: { minimumPaymentAmount: Big(acquiringIntegration.minimumPaymentAmount).toFixed(2) },
+                    }, context)
+                }
+
+                const payments = await Promise.all(paymentCreateInputs.map((paymentInput) => Payment.create(context, paymentInput)))
```
Comment (VKislov) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js:710]: same as above

**Review by VKislov — APPROVED** [review]

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.js (hunk 11 above)
Comment (VKislov) [reply apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.js:247]: oh no, everything is correct here. The operation is asynchronous, so there will always be a promise here, and PromiseAll works with promises and will wait for them to complete by default

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentForVirtualReceiptService.js (hunk 7 above)
Comment (dkoviazin) [reply apps/condo/domains/acquiring/schema/RegisterMultiPaymentForVirtualReceiptService.js:202]: Before multiPayment is created, several Payments are created. We can not check the total amount of MultiPayment inside Payment model. So, we can not implement hook to Payment model and if we first create payments and then throw error in MultiPayment - all created Payments will be useless

File: apps/condo/lang/ru/ru.json (hunk 8 above)
Comment (dkoviazin) [reply apps/condo/lang/ru/ru.json:3137]: It's hard to fix translations from another PR. No context. And texts were approved by product owners

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js (hunk 10 above)
Comment (dkoviazin) [reply apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js:708]: we need to remove copy-paste from all mutations. It is not the goal of this PR

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js (hunk 9 above)
Comment (dkoviazin) [reply apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js:357]: When user want to make a partial payment he needs to pass amounts for every payment he choose. Agreement with mobile APP

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.js (hunk 4 above)
Comment (dkoviazin) [reply apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.js:243]: This is not the only place for rounding numbers to 2 digits. If we will use BTC - we will need a task to fix all places. For now, we are working only with FIAT currency

Comment (pahaz) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.js:243]: Why to fix 2? It’s money related magic const

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentForOneReceiptService.js [hunk 14]
```diff
@@ -241,6 +245,15 @@ const RegisterMultiPaymentForOneReceiptService = new GQLCustomSchema('RegisterMu
                     implicitFee: String(implicitFee),
                     serviceFee: String(fromReceiptAmountFee),
                 }
+                const amountToPay = Big(billingReceipt.toPay)
+                    .add(Big(paymentCommissionFields.explicitServiceCharge))
+                    .add(Big(paymentCommissionFields.explicitFee))
+                if (acquiringIntegration.minimumPaymentAmount && Big(amountToPay).lt(acquiringIntegration.minimumPaymentAmount)) {
+                    throw new GQLError({
+                        ...ERRORS.PAYMENT_AMOUNT_LESS_THAN_MINIMUM,
+                        messageInterpolation: { minimumPaymentAmount: Big(acquiringIntegration.minimumPaymentAmount).toFixed(2) },
```
Comment (pahaz) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentForOneReceiptService.js:254]: Please remove toFixed. It’s local related. It’s better to leave it as is

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentForVirtualReceiptService.js [hunk 15]
```diff
@@ -185,6 +189,17 @@ const RegisterMultiPaymentForVirtualReceiptService = new GQLCustomSchema('Regist
...
                     serviceFee: String(fromReceiptAmountFee),
                 }
+
+                const amountToPay = Big(amount)
+                    .add(Big(paymentCommissionFields.explicitServiceCharge))
+                    .add(Big(paymentCommissionFields.explicitFee))
+                if (acquiringIntegration.minimumPaymentAmount && Big(amountToPay).lt(acquiringIntegration.minimumPaymentAmount)) {
+                    throw new GQLError({
+                        ...ERRORS.PAYMENT_AMOUNT_LESS_THAN_MINIMUM,
+                        messageInterpolation: { minimumPaymentAmount: Big(acquiringIntegration.minimumPaymentAmount).toFixed(2) },
```
Comment (pahaz) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentForVirtualReceiptService.js:199]: To fixed 2 is locale and money specific value

File: apps/condo/migrations/20240829153509-0418_acquiringintegration_minimumpaymentamount_and_more.js [hunk 16]
```diff
@@ -0,0 +1,34 @@
+// auto generated by kmigrator
+// KMIGRATOR:0418_acquiringintegration_minimumpaymentamount_and_more:IyBHZW5lcmF0ZWQgYnkgRGphbmdvIDQuMi4xNSBvbiAyMDI0LTA4LTI5IDEwOjM1Cgpmcm9tIGRqYW5nby5kYiBpbXBvcnQgbWlncmF0aW9ucywgbW9kZWxzCgoKY2xhc3MgTWlncmF0aW9uKG1pZ3JhdGlvbnMuTWlncmF0aW9uKToKCiAgICBkZXBlbmRlbmNpZXMgPSBbCiAgICAgICAgKCdfZGphbmdvX3NjaGVtYScsICcwNDE3X2JpbGxpbmdyZWNlaXB0X2JpbGxpbmdyZWNlaXB0X3BlcmlvZF9jb250ZXh0JyksCiAgICBdCgogICAgb3BlcmF0aW9ucyA9IFsKICAgICAgICBtaWdyYXRpb25zLkFkZEZpZWxkKAogICAgICAgICAgICBtb2RlbF9uYW1lPSdhY3F1aXJpbmdpbnRlZ3JhdGlvbicsCiAgICAgICAgICAgIG5hbWU9J21pbmltdW1QYXltZW50QW1vdW50JywKICAgICAgICAgICAgZmllbGQ9bW9kZWxzLkRlY2ltYWxGaWVsZChibGFuaz1UcnVlLCBkZWNpbWFsX3BsYWNlcz04LCBtYXhfZGlnaXRzPTE4LCBudWxsPVRydWUpLAogICAgICAgICksCiAgICAgICAgbWlncmF0aW9ucy5BZGRGaWVsZCgKICAgICAgICAgICAgbW9kZWxfbmFtZT0nYWNxdWlyaW5naW50ZWdyYXRpb25oaXN0b3J5cmVjb3JkJywKICAgICAgICAgICAgbmFtZT0nbWluaW11bVBheW1lbnRBbW91bnQnLAogICAgICAgICAgICBmaWVsZD1tb2RlbHMuRGVjaW1hbEZpZWxkKGJsYW5rPVRydWUsIGRlY2ltYWxfcGxhY2VzPTQsIG1heF9kaWdpdHM9MTgsIG51bGw9VHJ1ZSksCiAgICAgICAgKSwKICAgIF0K
+
+exports.up = async (knex) => {
+    await knex.raw(`
+    BEGIN;
+--
```
Comment (pahaz) [inline apps/condo/migrations/20240829153509-0418_acquiringintegration_minimumpaymentamount_and_more.js:7]: You must increase Timors for this changes

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js [hunk 17]
```diff
@@ -553,88 +556,83 @@ const RegisterMultiPaymentService = new GQLCustomSchema('RegisterMultiPaymentSer
                 }
 
                 // Stage 3 Generating payments
-                const payments = []
```
Comment (pahaz) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js:556]: It looks like too many changes for your task. Why are you rewrite stage 3 ? @dkoviazin could you please check this changes

Comment (dkoviazin) [reply apps/condo/domains/acquiring/schema/RegisterMultiPaymentService.js:556]: It was a copy-paste from the start. 
We have many tests for registerMultiPayment with invoices or receipts and they are all working
So, LGTM

File: apps/condo/domains/acquiring/constants/errors.js [hunk 18]
```diff
@@ -2,6 +2,8 @@ const { GQLError, GQLErrorCode: { BAD_USER_INPUT } } = require('@open-condo/keys
 
 const { PAYMENT_DONE_STATUS, MULTIPAYMENT_DONE_STATUS, PAYMENT_INIT_STATUS } = require('./payment')
 
+const { WRONG_VALUE } = require('../../common/constants/errors')
```
Comment (SavelevMatthew) [inline apps/condo/domains/acquiring/constants/errors.js:5]: Use `@condo/domain` imports

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.test.js [hunk 19]
```diff
@@ -277,4 +282,62 @@ describe('RegisterMultiPaymentForInvoicesService', () => {
             ]))
         })
     })
+
+    describe('RegisterMultiPaymentForInvoicesService minimum amount check', () => {
```
Comment (SavelevMatthew) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.test.js:286]: This block should be inside validations, shouldn't it?

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.test.js [hunk 20]
```diff
@@ -277,4 +282,62 @@ describe('RegisterMultiPaymentForInvoicesService', () => {
...
+            })
+        })
+
+        afterEach(async () => {
+            await utils.updateAcquiringIntegration({ minimumPaymentAmount: null })
+        })
+
+        describe('Check minimum payment amount from acquiring integration', () => {
+            test('Payment for acquiring with no set the minimum payment amount', async () => {
+                const [invoice] = await createTestInvoice(utils.clients.admin, utils.organization, { status: INVOICE_STATUS_PUBLISHED })
```
Comment (SavelevMatthew) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentForInvoicesService.test.js:309]: Why admins are everywhere?

File: apps/condo/domains/acquiring/schema/RegisterMultiPaymentForOneReceiptService.test.js [hunk 21]
```diff
@@ -34,6 +36,15 @@ const {
     createTestBillingIntegration,
     updateTestBillingIntegrationOrganizationContext,
 } = require('@condo/domains/billing/utils/testSchema')
+const {
+    TestUtils,
+    ResidentTestMixin,
+} = require('@condo/domains/billing/utils/testSchema/testUtils')
+
+const { INVOICE_STATUS_PUBLISHED } = require('../../marketplace/constants')
```
Comment (SavelevMatthew) [inline apps/condo/domains/acquiring/schema/RegisterMultiPaymentForOneReceiptService.test.js:44]: imports!

**Pull request merged**